querqy-core/src/main/java/querqy/rewrite/RewriteChain.java

```diff
@@ -24,62 +23,112 @@ public class RewriteChain {
+
+        private final SearchEngineRequestAdapter searchEngineRequestAdapter;
+        private final RewriteLoggingConfig rewriteLoggingConfig;
 
-        final String oldRewriterId = loggingContext.map(InfoLoggingContext::getRewriterId).orElse(null);
+        private ExpandedQuery expandedQuery;
 
-        try {
+        private final RewriteChainLog.RewriteChainLogBuilder rewriteChainLoggingBuilder = RewriteChainLog.builder();
 
-            for (final RewriterFactory factory : factories) {
+        public RewritingExecutor(
+                final List<RewriterFactory> rewriterFactories,
+                final SearchEngineRequestAdapter searchEngineRequestAdapter,
+                final ExpandedQuery expandedQuery
+        ) {
+            this.rewriterFactories = rewriterFactories;
+
+            this.searchEngineRequestAdapter = searchEngineRequestAdapter;
+            this.rewriteLoggingConfig = searchEngineRequestAdapter.getRewriteLoggingConfig();
+
+            this.expandedQuery = expandedQuery;
+        }
 
-                loggingContext.ifPresent(context -> context.setRewriterId(factory.getRewriterId()));
+        public RewriteChainOutput rewrite() {
+            for (final RewriterFactory factory : rewriterFactories) {
+                final RewriterOutput rewriterOutput = applyFactory(factory);
 
-                final QueryRewriter rewriter = factory.createRewriter(
-                        rewrittenQuery.getExpandedQuery(), searchEngineRequestAdapter);
+                if (rewriteLoggingConfig.isActive() && rewriterOutput.getRewriterLogging().isPresent()) {
+                    addLoggingIfRewritingHasBeenApplied(
+                            factory.getRewriterId(), rewriterOutput.getRewriterLogging().get());
+                }
 
-                rewrittenQuery = rewriter.rewrite(rewrittenQuery.getExpandedQuery(), searchEngineRequestAdapter);
+                expandedQuery = rewriterOutput.getExpandedQuery();
             }
 
-        } finally {
-            loggingContext.ifPresent(context -> context.setRewriterId(oldRewriterId));
+            return buildOutput();
         }
 
-        return rewrittenQuery.getExpandedQuery();
-    }
+        private RewriterOutput applyFactory(final RewriterFactory factory) {
+            final QueryRewriter rewriter = factory.createRewriter(expandedQuery, searchEngineRequestAdapter);
+            return rewriter.rewrite(expandedQuery, searchEngineRequestAdapter);
+        }
 
-    @Deprecated
-    public List<RewriterFactory> getRewriterFactories() {
-        return factories;
-    }
+        private void addLoggingIfRewritingHasBeenApplied(final String factoryId, final RewriterLog rewriterLogging) {
+            if (rewriterLogging.hasAppliedRewriting()) {
+                addLoggingIfRewriterIdIsIncluded(factoryId, rewriterLogging);
+            }
+        }
+
+        private void addLoggingIfRewriterIdIsIncluded(final String factoryId, final RewriterLog rewriterLogging) {
```

Comment:
I'd prefer to not interpret includedRewriters.empty()=true as 'logging enabled for all rewriters. As soon as someone turns on one rewriter, suddenly all others become disabled. In ES we required each rewriter that is to be logged to have logging enabled - that should be the most robust option. I'd also prefer renaming `included` to something `loggingEnabled` (see querqy.elasticsearch.RewriteChainAndLogging#rewritersEnabledForLogging)

Reply:
I think it is quite counterintuitive to set the logging as enabled and not to get anything back. Furthermore, I think it is not counterintuitive to return logging only for one rewriter if only one rewriter is specified. I do not see the use case of having logging enabled while activating no rewriter.

Reply:
Following our conversation, we will tackle this when we refactor the Sink concept in a follow-up step